include/nbl/asset/filters/CNormalMapToDerivativeFilter.h

```diff
@@ -0,0 +1,275 @@
+
+						auto* inDataAdress = inData + readBlockArrayOffset;
+						const void* inSourcePixels[maxPlanes] = { inDataAdress, nullptr, nullptr, nullptr };
+
+						decodeType decodeBuffer[maxChannels] = {};
+						for (auto blockY = 0u; blockY < blockDims.y; blockY++)
+							for (auto blockX = 0u; blockX < blockDims.x; blockX++)
+							{
+								asset::decodePixelsRuntime(inFormat, inSourcePixels, decodeBuffer, blockX, blockY);
+								const size_t offset = asset::IImage::SBufferCopy::getLocalByteOffset(core::vector3du32_SIMD(localOutPos.x + blockX, localOutPos.y + blockY, localOutPos.z), scratchByteStrides);
+								memcpy(reinterpret_cast<uint8_t*>(scratchMemory) + offset, decodeBuffer, scratchTexelByteSize);
+							}
+					};
+
+					IImage::SSubresourceLayers subresource = { static_cast<IImage::E_ASPECT_FLAGS>(0u), state->inMipLevel, state->inBaseLayer, 1 };
+					CMatchedSizeInOutImageFilterCommon::state_type::TexelRange range = { state->inOffset,state->extent };
+					CBasicImageFilterCommon::clip_region_functor_t clipFunctor(subresource, range, inFormat);
+
+					auto& inRegions = state->inImage->getRegions(state->inMipLevel);
+					CBasicImageFilterCommon::executePerRegion(state->inImage, decode, inRegions.begin(), inRegions.end(), clipFunctor);
+				}
+
+				{
+					auto getScratchPixel = [&](core::vector4di32_SIMD readBlockPos) -> decodeType*
+					{
+						const size_t scratchOffset = asset::IImage::SBufferCopy::getLocalByteOffset(core::vector3du32_SIMD(readBlockPos.x, readBlockPos.y, readBlockPos.z, 0), scratchByteStrides);
+						return reinterpret_cast<decodeType*>(reinterpret_cast<uint8_t*>(scratchMemory) + scratchOffset);
+					};
+
+					auto computeDerivativeTexel = [&](core::vectorSIMDi32 readBlockPos) -> void
+					{
+						decodeType* current = getScratchPixel(readBlockPos);
+						auto& [x, y, z, a] = std::make_tuple(*current, *(current + 1), *(current + 2), *(current + 3));
+
+						std::for_each(current, current + currentChannelCount,
+							[&](const decodeType& itrValue)
+							{
+								uint8_t offset = &itrValue - current;
+								const decodeType absoluteValue = core::abs(itrValue);
+							
+								if (maxAbsoluteDecodeValues[offset] < absoluteValue)
+									maxAbsoluteDecodeValues[offset] = absoluteValue;
+							}
+						);
+
+						x = -x / z;
+						y = -y / z;
+					};
+
+					{
+						core::vector3du32_SIMD localCoord;
+						for (auto& z = localCoord[2] = 0u; z < state->extent.depth; ++z)
+							for (auto& y = localCoord[1] = 0u; y < state->extent.height; ++y)
+								for (auto& x = localCoord[0] = 0u; x < state->extent.width; ++x)
+									computeDerivativeTexel(core::vectorSIMDu32(x, y, z));
+					}
+
+					auto& maxAbsLayerScaleValues = state->maxAbsLayerScaleValues.emplace_back();
+					for (auto& absLayerScaleValue : maxAbsLayerScaleValues)
+						absLayerScaleValue = maxAbsoluteDecodeValues[&absLayerScaleValue - &maxAbsLayerScaleValues[0]];
```

Comment:
it would be far better to carve out some memory from the state's scratch and use that instead of a dynamically allocated vector

Reply:
corrected in following commit